cargo-dist/src/build/mod.rs

```diff
@@ -0,0 +1,229 @@
+pub struct BuildExpectations {
+    /// Expectations grouped by package
+    pub packages: SortedMap<String, BinaryExpectations>,
+    /// Whether this is fake (--artifacts=lies)
+    fake: bool,
+}
+
+/// Output expectations for builds, and computed facts (one package)
+#[derive(Default)]
+pub struct BinaryExpectations {
+    /// Expected binaries
+    pub binaries: SortedMap<String, ExpectedBinary>,
+}
+
+/// Binaries we expect
+pub struct ExpectedBinary {
+    /// idx of the binary in the DistGraph
+    pub idx: BinaryIdx,
+    /// path to the binary in the build output
+    ///
+    /// Initially this is None, but should be Some by the end of the build from calls to found_bin
+    pub src_path: Option<Utf8PathBuf>,
+    /// paths to the symbols of this binary in the build output
+    ///
+    /// Initially this is empty, but should be Some by the end of the build from calls to found_bin
+    pub sym_paths: Vec<Utf8PathBuf>,
+}
+
+impl BuildExpectations {
+    /// Create a new BuildExpectations
+    pub fn new(dist: &DistGraph, expected_binaries: &[BinaryIdx]) -> Self {
+        let mut packages = SortedMap::<String, BinaryExpectations>::new();
+        for &binary_idx in expected_binaries {
+            let binary = &dist.binary(binary_idx);
+
+            // Get the package id or an empty string (for generic builds)
+            let package_id = package_id_string(binary.pkg_id.as_ref());
+            let exe_name = binary.name.clone();
+
+            packages.entry(package_id).or_default().binaries.insert(
+                exe_name,
+                ExpectedBinary {
+                    idx: binary_idx,
+                    src_path: None,
+                    sym_paths: vec![],
+                },
+            );
+        }
+
+        Self {
+            packages,
+            fake: false,
+        }
+    }
+
+    /// Create a new BuildExpectations, but don't sweat things being faked
+    ///
+    /// This is used for --artifacts=lies
+    pub fn new_fake(dist: &DistGraph, expected_binaries: &[BinaryIdx]) -> Self {
+        let mut out = Self::new(dist, expected_binaries);
+        out.fake = true;
+        out
+    }
+
+    /// Report that a binary was found, which might have been expected
+    ///
+    /// This subroutine is responsible for sorting out whether we care about the binary,
+    /// and if the maybe_symbols are in fact symbols we care about.
+    pub fn found_bin(
+        &mut self,
+        pkg_id: String,
+        src_path: Utf8PathBuf,
+        maybe_symbols: Vec<Utf8PathBuf>,
+    ) {
+        info!("got a new binary: {}", src_path);
+
+        // lookup the package
+        let Some(pkg) = self.packages.get_mut(&pkg_id) else {
+            return;
+        };
+
+        // lookup the binary in the package
+        let Some(bin_name) = src_path.file_stem() else {
+            return;
+        };
+        let Some(bin_result) = pkg.binaries.get_mut(bin_name) else {
+            return;
```

Comment:
I like separating out/deduplicating this logic! Much cleaner.

That said, I feel like we should add an error case for if the binary it's been passed doesn't exist in the expected set it's looking at. When this logic was centralized in one place, we could be pretty sure we wouldn't end up with "bad" unexpected binaries; now that this is a method that could hypothetically get called from anywhere, that logic error is easier to make in the future.

Reply:
So it's actually Intended Design that this API gets handed a bunch of junk paths, because this API is specifically processing the output from rustc/cargo, which reports tons and tons of intermediate/uninteresting results.

Previously it was the job of build_cargo to filter those out, but I specifically moved the logic into the generic driver so that it could be more "head empty", and so other future backends could potentially similarly be like "i see a bunch of binaries in this dir, did we need any of these, oh wise BuildExpectations?"

Reply:
That makes sense; can we document that somewhere explicitly? I think I'm going to forget this and become surprised again in the future if this gets handed a bunch of unexpected files.